src/java.base/share/classes/java/security/Provider.java

```diff
@@ -913,149 +913,157 @@ private void implPutAll(Map<?,?> t) {
     }
 
     private Object implRemove(Object key) {
-        if (key instanceof String) {
-            if (!checkLegacy(key)) {
-                return null;
-            }
-            legacyStrings.remove((String)key);
+        if (!checkLegacy(key)) return null;
+
+        Object o = super.remove(key);
+        if (o != null && o instanceof String && key instanceof String) {
```

Comment:
`o != null` seems redundant here. `o instanceof String` will be false in case of `o == null`

Reply:
Yup, removed.